### PR TITLE
DPB-ACL scale tests (#30)

### PR DESCRIPTION
**Pull request by vasant17**
* DPB-ACL scale tests and code-review comments addressed

* Remove sleeps that were put earlier

Co-authored-by: Vasant <vapatil@linkedin.com>

**What I did**
Wrote DPB scale test cases
**Why I did it**
To catch any race conditions or corner cases

**How I verified it**
Ran VS test cases

**Details if related**
```
vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$ sudo pytest --pdb -s -v --dvsname=vs-vp test_port_dpb_acl.py
======================================================================= test session starts ========================================================================
platform linux2 -- Python 2.7.15+, pytest-3.3.0, py-1.8.0, pluggy-0.6.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/vapatil/workspace/DPB/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 5 items

test_port_dpb_acl.py::TestPortDPBAcl::test_acl_table_empty_port_list remove extra link dummy
Exception AssertionError: AssertionError(u"assert 0 > 0\n +  where 0 = int('0')\n +    where '0' = <built-in method group of _sre.SRE_Match object at 0x7f3062bcde40>(1)\n +      where <built-in method group of _sre.SRE_Match object at 0x7f3062bcde40> = <_sre.SRE_Match object at 0x7f3062bcde40>.group",) in <bound method ApplDbValidator.__del__ of <conftest.ApplDbValidator object at 0x7f3062bcbf10>> ignored
PASSED                                                                                  [ 20%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_two_acl_tables PASSED                                                                                    [ 40%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_acl_table_many_ports PASSED                                                                                   [ 60%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_many_acl_tables PASSED                                                                                   [ 80%]
test_port_dpb_acl.py::TestPortDPBAcl::test_all_ports_two_acl_tables PASSED                                                                                   [100%]Exception AssertionError: AssertionError(u"assert 0 > 0\n +  where 0 = int('0')\n +    where '0' = <built-in method group of _sre.SRE_Match object at 0x7f3062caa648>(1)\n +      where <built-in method group of _sre.SRE_Match object at 0x7f3062caa648> = <_sre.SRE_Match object at 0x7f3062caa648>.group",) in <bound method ApplDbValidator.__del__ of <conftest.ApplDbValidator object at 0x7f3062c15f10>> ignored


=================================================================== 5 passed in 1080.87 seconds ====================================================================
vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$
```

Test cases after second commit
```
$ sudo pytest -s -v --dvsname=vs-vp test_acl.py test_port_dpb_acl.py
================================================================================== test session starts ==================================================================================
platform linux2 -- Python 2.7.17, pytest-3.3.0, py-1.8.0, pluggy-0.6.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/vapatil/workspace/DPB/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 29 items

test_acl.py::TestAcl::test_AclTableCreation remove extra link dummy
PASSED                                                                                                                                [  3%]
test_acl.py::TestAcl::test_AclRuleL4SrcPort PASSED                                                                                                                                [  6%]
test_acl.py::TestAcl::test_AclRuleInOutPorts PASSED                                                                                                                               [ 10%]
test_acl.py::TestAcl::test_AclRuleInPortsNonExistingInterface PASSED                                                                                                              [ 13%]
test_acl.py::TestAcl::test_AclRuleOutPortsNonExistingInterface PASSED                                                                                                             [ 17%]
test_acl.py::TestAcl::test_AclTableDeletion PASSED                                                                                                                                [ 20%]
test_acl.py::TestAcl::test_V6AclTableCreation PASSED                                                                                                                              [ 24%]
test_acl.py::TestAcl::test_V6AclRuleIPv6Any PASSED                                                                                                                                [ 27%]
test_acl.py::TestAcl::test_V6AclRuleIPv6AnyDrop PASSED                                                                                                                            [ 31%]
test_acl.py::TestAcl::test_V6AclRuleIpProtocol PASSED                                                                                                                             [ 34%]
test_acl.py::TestAcl::test_V6AclRuleSrcIPv6 PASSED                                                                                                                                [ 37%]
test_acl.py::TestAcl::test_V6AclRuleDstIPv6 PASSED                                                                                                                                [ 41%]
test_acl.py::TestAcl::test_V6AclRuleL4SrcPort PASSED                                                                                                                              [ 44%]
test_acl.py::TestAcl::test_V6AclRuleL4DstPort PASSED                                                                                                                              [ 48%]
test_acl.py::TestAcl::test_V6AclRuleTCPFlags PASSED                                                                                                                               [ 51%]
test_acl.py::TestAcl::test_V6AclRuleL4SrcPortRange PASSED                                                                                                                         [ 55%]
test_acl.py::TestAcl::test_V6AclRuleL4DstPortRange PASSED                                                                                                                         [ 58%]
test_acl.py::TestAcl::test_V6AclTableDeletion PASSED                                                                                                                              [ 62%]
test_acl.py::TestAcl::test_InsertAclRuleBetweenPriorities PASSED                                                                                                                  [ 65%]
test_acl.py::TestAcl::test_RulesWithDiffMaskLengths PASSED                                                                                                                        [ 68%]
test_acl.py::TestAcl::test_AclRuleIcmp PASSED                                                                                                                                     [ 72%]
test_acl.py::TestAcl::test_AclRuleIcmpV6 PASSED                                                                                                                                   [ 75%]
test_acl.py::TestAcl::test_AclRuleRedirectToNextHop PASSED                                                                                                                        [ 79%]
test_acl.py::TestAclRuleValidation::test_AclActionValidation PASSED                                                                                                               [ 82%]
test_port_dpb_acl.py::TestPortDPBAcl::test_acl_table_empty_port_list remove extra link dummy
PASSED                                                                                                       [ 86%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_two_acl_tables PASSED                                                                                                         [ 89%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_acl_table_many_ports PASSED                                                                                                        [ 93%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_many_acl_tables PASSED                                                                                                        [ 96%]
test_port_dpb_acl.py::TestPortDPBAcl::test_many_ports_many_acl_tables PASSED                                                                                                      [100%]

============================================================================== 29 passed in 454.67 seconds ==============================================================================

```

Test result after third commit
```
vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$ sudo pytest -s -v --pdb --dvsname=vs-vp test_port_dpb_acl.py
================================================================================== test session starts ==================================================================================
platform linux2 -- Python 2.7.17, pytest-3.3.0, py-1.8.0, pluggy-0.6.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/vapatil/workspace/DPB/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 5 items                                                                                                                                                                       

test_port_dpb_acl.py::TestPortDPBAcl::test_acl_table_empty_port_list remove extra link dummy
Exception AssertionError: AssertionError(u"assert 0 > 0\n +  where 0 = int('0')\n +    where '0' = <built-in method group of _sre.SRE_Match object at 0x7f520407f130>(1)\n +      where <built-in method group of _sre.SRE_Match object at 0x7f520407f130> = <_sre.SRE_Match object at 0x7f520407f130>.group",) in <bound method ApplDbValidator.__del__ of <conftest.ApplDbValidator object at 0x7f5203fd5090>> ignored
PASSED                                                                                                       [ 20%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_two_acl_tables PASSED                                                                                                         [ 40%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_acl_table_many_ports PASSED                                                                                                        [ 60%]
test_port_dpb_acl.py::TestPortDPBAcl::test_one_port_many_acl_tables PASSED                                                                                                        [ 80%]
test_port_dpb_acl.py::TestPortDPBAcl::test_many_ports_many_acl_tables PASSED                                                                                                      [100%]Exception AssertionError: AssertionError(u"assert 0 > 0\n +  where 0 = int('0')\n +    where '0' = <built-in method group of _sre.SRE_Match object at 0x7f52040457b0>(1)\n +      where <built-in method group of _sre.SRE_Match object at 0x7f52040457b0> = <_sre.SRE_Match object at 0x7f52040457b0>.group",) in <bound method ApplDbValidator.__del__ of <conftest.ApplDbValidator object at 0x7f5203f7fa90>> ignored


============================================================================== 5 passed in 332.76 seconds ===============================================================================
vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$

vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$ sudo pytest -s -v --pdb --dvsname=vs-vp test_acl.py
================================================================================== test session starts ==================================================================================
platform linux2 -- Python 2.7.17, pytest-3.3.0, py-1.8.0, pluggy-0.6.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/vapatil/workspace/DPB/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 24 items

test_acl.py::TestAcl::test_AclTableCreation remove extra link dummy
PASSED                                                                                                                                [  4%]
test_acl.py::TestAcl::test_AclRuleL4SrcPort PASSED                                                                                                                                [  8%]
test_acl.py::TestAcl::test_AclRuleInOutPorts PASSED                                                                                                                               [ 12%]
test_acl.py::TestAcl::test_AclRuleInPortsNonExistingInterface PASSED                                                                                                              [ 16%]
test_acl.py::TestAcl::test_AclRuleOutPortsNonExistingInterface PASSED                                                                                                             [ 20%]
test_acl.py::TestAcl::test_AclTableDeletion PASSED                                                                                                                                [ 25%]
test_acl.py::TestAcl::test_V6AclTableCreation PASSED                                                                                                                              [ 29%]
test_acl.py::TestAcl::test_V6AclRuleIPv6Any PASSED                                                                                                                                [ 33%]
test_acl.py::TestAcl::test_V6AclRuleIPv6AnyDrop PASSED                                                                                                                            [ 37%]
test_acl.py::TestAcl::test_V6AclRuleIpProtocol PASSED                                                                                                                             [ 41%]
test_acl.py::TestAcl::test_V6AclRuleSrcIPv6 PASSED                                                                                                                                [ 45%]
test_acl.py::TestAcl::test_V6AclRuleDstIPv6 PASSED                                                                                                                                [ 50%]
test_acl.py::TestAcl::test_V6AclRuleL4SrcPort PASSED                                                                                                                              [ 54%]
test_acl.py::TestAcl::test_V6AclRuleL4DstPort PASSED                                                                                                                              [ 58%]
test_acl.py::TestAcl::test_V6AclRuleTCPFlags PASSED                                                                                                                               [ 62%]
test_acl.py::TestAcl::test_V6AclRuleL4SrcPortRange PASSED                                                                                                                         [ 66%]
test_acl.py::TestAcl::test_V6AclRuleL4DstPortRange PASSED                                                                                                                         [ 70%]
test_acl.py::TestAcl::test_V6AclTableDeletion PASSED                                                                                                                              [ 75%]
test_acl.py::TestAcl::test_InsertAclRuleBetweenPriorities PASSED                                                                                                                  [ 79%]
test_acl.py::TestAcl::test_RulesWithDiffMaskLengths PASSED                                                                                                                        [ 83%]
test_acl.py::TestAcl::test_AclRuleIcmp PASSED                                                                                                                                     [ 87%]
test_acl.py::TestAcl::test_AclRuleIcmpV6 PASSED                                                                                                                                   [ 91%]
test_acl.py::TestAcl::test_AclRuleRedirectToNextHop PASSED                                                                                                                        [ 95%]
test_acl.py::TestAclRuleValidation::test_AclActionValidation PASSED                                                                                                               [100%]Exception AssertionError: AssertionError(u"assert 0 > 0\n +  where 0 = int('0')\n +    where '0' = <built-in method group of _sre.SRE_Match object at 0x7efdf26ac730>(1)\n +      where <built-in method group of _sre.SRE_Match object at 0x7efdf26ac730> = <_sre.SRE_Match object at 0x7efdf26ac730>.group",) in <bound method ApplDbValidator.__del__ of <conftest.ApplDbValidator object at 0x7efdf234b250>> ignored


============================================================================== 24 passed in 140.27 seconds ==============================================================================
vapatil@server09:~/workspace/DPB/sonic-buildimage/src/sonic-swss/tests$
```

### CODE REVIEW

**Comment by lguohan** [comment]
please use the new assert api.

File: tests/test_acl.py [hunk 1]
```diff
@@ -1,417 +1,240 @@
...
-                    continue
-                elif k == "SAI_ACL_RANGE_ATTR_LIMIT" and v == expected_ports:
-                    continue
-                else:
-                    return False
-
-            return True
-
-        return _match_acl_range
+    def setup_dvs_acl(self, dvs):
```
Comment (daall) [inline tests/test_acl.py:4]: I think we can probably get rid of BaseTestAcl at this point since we've moved all the helpers and just provide a fixture to access dvs_acl.

File: tests/dvslib/dvs_acl.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+        acl_tables = [k for k in keys if k not in self.asic_db.default_acl_tables]
+
+        return acl_tables
+
+    def get_acl_table_id(self):
+        acl_tables = self.get_acl_table_ids()
+        return acl_tables[0]
+
+    def verify_acl_tables(self, expt):
```
[the remark this answers is not in the document]
Comment (vasant17) [reply tests/dvslib/dvs_acl.py:50]: Yes.

File: tests/test_acl.py (hunk 1 above)
Comment (vasant17) [reply tests/test_acl.py:4]: 👍

File: tests/conftest.py [hunk 3]
```diff
@@ -953,31 +956,40 @@ def get_asic_db(self):
...
         return self.flex_db
-    
+
     def get_state_db(self):
         if not self.state_db:
             self.state_db = dvs_db.DVSDatabase(self.STATE_DB_ID, self.redis_sock)
 
         return self.state_db
 
+    def get_dvs_acl(self):
```
[the remark this answers is not in the document]
Comment (vasant17) [reply tests/conftest.py:984]: 👍

File: tests/conftest.py [hunk 4]
```diff
@@ -998,6 +1009,13 @@ def testlog(request, dvs):
     yield testlog
     dvs.runcmd("logger === finish test %s ===" % request.node.name)
 
+@pytest.yield_fixture(scope="class")
```
Comment (daall) [inline tests/conftest.py:1012]: I don't think it makes a difference for our purposes but we should probably keep this at the same scope as dvs since it depends on dvs.

Comment (vasant17) [reply tests/conftest.py:1012]: I do NOT want to call dvs_acl_manager fixture per method. I want it to be called once per class. Why do you think it should be at the same scope as that of dvs? I think scope for fixtures should be derived by need.

Comment (daall) [reply tests/conftest.py:1012]: I get what you're saying now! Let's go ahead and keep it as is then.
